### ISSUE TITLE
Better interactivity in low-memory situations on Linux desktop

### ISSUE THREAD

**Issue by hakavlad** [ninja-build/ninja#1706]
Hi! There is one problem when using ninja on the desktop: when compiling large projects in low memory conditions with active swapping, the desktop may completely freeze.

Fedora Worksation guys discuss ways to improve desktop interactivity in low-memory situations. A typical case that causes a state of low memory, poor interactivity, and even complete freezing is a compilation of large projects: https://pagure.io/fedora-workstation/issue/98#comment-588713.

It seems that there is already an option to increase interactivity:

```
  -l N     do not start new jobs if the load average is greater than N
```

I suggest also using options based on the use of `PSI` in addition to the `-l` option. PSI is a new metrics that appeared in Linux 4.20.

About PSI:

- https://lwn.net/Articles/783520/
- https://lwn.net/Articles/759658/
- https://lwn.net/Articles/759781/
- https://facebookmicrosites.github.io/psi/
- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/accounting/psi.rst
- https://www.kernel.org/doc/html/latest/accounting/psi.html
- https://unixism.net/2019/08/linux-pressure-stall-information-psi-by-example/

PSI metrics allow accurate estimation of io and memory pressure. My suggestion is to add new CLI options:

```
  --psi-mem N   do not start new jobs if PSI memory pressure some_avg10 is greater than N
  --psi-io N    do not start new jobs if PSI io pressure some_avg10 is greater than N
```

I would like to run a ninja with the options `--psi-mem 20` and `--psi-io 40` so that with low-memory conditions the minimum responsiveness of the desktop is maintained.

This should not be difficult to implement - the necessary numbers can be taken from the files /proc/pressure/memory and /proc/pressure/io.

**Comment by scivision** [comment]
I have noticed this RAM issue on multi-core embedded systems.  A popular example is the 1GB Raspberry Pi 4, it's easy to overwhelm the RAM.

**Comment by scivision** [comment]
One common example of this being an issue is building CMake using Ninja--the later parts of the build easily overwhelm a 16 GB RAM PC. I have worked around the issue of overwhelmed RAM on my own projects by setting up Ninja job pools in CMakeLists.txt, but this is a manual, iterative, perhaps last-resort approach. I am glad that job pools exist because the whole build would be slowed drastically otherwise.

**Comment by advancedwebdeveloper** [comment]
I think I have a different situation.
I have 8GB of physical RAM. Around 1,5 GB are used by Ubuntu 20 x86_64 (the kernel, drivers and the userspace). The rest is available for my needs.
I do have around 3GB of swap mounted to an external SSD drive.
While not all swap is used - I am receiving "out of memory" error and immediate exit of Clang.
I am planning to track down memory allocator appetites.
It could be interesting to run some of Ninja's sub-tasks in different order, if those do not depend on each other. That could allow to identify specific source files, in C++, which require too much RAM to compile.

**Comment by scivision** [comment]
Yes I think the situation you're describing is more severe in that errors are occurring--I get similar errors as well sometimes. This is also an issue when using GNU `make -j` instead of Ninja. I suppose unity builds could exacerbate this as well.

A question might be is this memory management for compilers in the scope of Ninja, or in the scope of the meta-build system (such as MSVC or CMake) driving Ninja. That is how I workaround these issues manually for bigger projects, by defining Ninja job pools manually in CMake for targets I know consume a lot of build resources (RAM). Your idea is more advanced but it would depend on if Ninja maintainers determine it to be in scope of Ninja design.

**Comment by ericoporto** [comment]
I have problems with ninja in Linux VMs/Containers in CI systems, which are giving me rather unpredictable memory requirements, it SOMETIMES causes OOM issues but not always, I can sometimes rerun in the same condition and get it to build, and sometimes not.

**Comment by ishitatsuyuki** [comment]
FWIW, I've tried implementing PSI-based spawn heuristics, and while it's an improvement over load average in reaction speed, there's still the problem that as soon as the PSI value falls under the threshold, Ninja will spawn process up to the `-j` limit transiently, which will bring back the system into a memory-pressured state again.

For better scheduling decisions we probably need some in-depth modeling over the compiler memory usage, mainly predicting the memory usage spike during compiler startup, so that only a "safe" amount of process is spawned at once and more process will only be spawned as we confirm the existing process won't use too much memory.

**Comment by scivision** [comment]
Sounds like what a Kalman filter might be able to do--not sure if one has been applied to memory usage before

**Comment by justanhduc** [comment]
Hello. Is there any solution to prevent this problem so far?

**Comment by hakavlad** [comment]
@justanhduc

>any solution to prevent this problem so far?

yes: https://github.com/hakavlad/nohang#solution
